src/fluree/db/query/fql/parse.cljc

```diff
@@ -0,0 +1,428 @@
+                      (throw (ex-info (str "Filter function must contain a valid variable. Provided: " code-str)
+                                      {:status 400 :error :db/invalid-query})))
+        var-name  (find-filtered-var code-vars vars)
+        params    (vec code-vars)
+        [fun _]   (filter/extract-filter-fn code code-vars)
+        f         (filter/make-executable params fun)]
+    (where/->function var-name params f)))
+
+(def ^:const default-recursion-depth 100)
+
+(defn recursion-predicate
+  "A predicate that ends in a '+', or a '+' with some integer afterwards is a recursion
+  predicate. e.g.: person/follows+3
+
+  Returns a two-tuple of predicate followed by # of times to recur.
+
+  If not a recursion predicate, returns nil."
+  [predicate context]
+  (when (or (string? predicate)
+            (keyword? predicate))
+    (when-let [[_ pred recur-n] (re-find #"(.+)\+(\d+)?$" (name predicate))]
+      [(json-ld/expand (keyword (namespace predicate) pred)
+                       context)
+       (if recur-n
+         (util/str->int recur-n)
+         default-recursion-depth)])))
+
+(defn parse-pred-ident
+  [x]
+  (when (util/pred-ident? x)
+    (where/->ident x)))
+
+(defn parse-iri
+  [x context]
+  (-> x
+      (json-ld/expand-iri context)
+      where/->value
+      (assoc ::where/datatype const/$xsd:anyURI)))
+
+(defn parse-sid
+  [x]
+  (when (syntax/sid? x)
+    (where/->value x)))
+
+(defn parse-subject
+  ([x]
+   (parse-sid x))
+
+  ([x context]
+   (if-let [parsed (parse-subject x)]
+     parsed
+     (when context
+       (parse-iri x context)))))
+
+(defn parse-subject-pattern
+  [s-pat context]
+  (when s-pat
+    (or (parse-variable s-pat)
+        (parse-pred-ident s-pat)
+        (parse-subject s-pat context)
+        (throw (ex-info (str "Subject values in where statement must be integer subject IDs or two-tuple identies. "
+                             "Provided: " s-pat ".")
+                        {:status 400 :error :db/invalid-query})))))
+
+(defn parse-class-predicate
+  [x]
+  (when (rdf-type? x)
+    (where/->value const/$rdf:type)))
+
+(defn parse-iri-predicate
+  [x]
+  (when (= "@id" x)
+    (where/->predicate const/$iri)))
+
+(defn iri->pred-id
+  [iri db context]
+  (let [full-iri (json-ld/expand-iri iri context)]
+    (dbproto/-p-prop db :id full-iri)))
+
+(defn iri->pred-id-strict
+  [iri db context]
+  (or (iri->pred-id iri db context)
+      (throw (ex-info (str "Invalid predicate: " iri)
+                      {:status 400 :error :db/invalid-query}))))
+
+(defn parse-recursion-predicate
+  [x db context]
+  (when-let [[p-iri recur-n] (recursion-predicate x context)]
+    (let [iri     (iri->pred-id-strict p-iri db context)
+          recur-n (or recur-n util/max-integer)]
+      (where/->predicate iri recur-n))))
+
+(defn parse-full-text-predicate
+  [x db context]
+  (when (and (string? x)
+             (str/starts-with? x "fullText:"))
+    (-> x
+        (subs 9)
```

Comment:
Looking through this PR, I've had this vague urge to suggest we add some tests for some of the helpers/internal representation steps. (Not as a blocker to this PR, necessarily, but ✨ someday✨ perhaps?) 

Parsing is implicitly tested by the query integration tests, but personally I'd like to have some tests for the pre-execution steps (validation, parsing). This new parsing pipeline is _much_ simpler than the old one, so the need for tests doesn't feel as urgent to me, but I still think it'd be valuable to document what this code is meant to output. 

While there's not _too_ much string-munging, regex, or magic number usage (this `9` isn't all that magical I guess), there is enough in here that it adds to my desire for tests. 😅 

Again, not a blocker for this PR, but something I've been thinking about.

Reply:
I agree. I think the internal parsed query representation is ready to be standardized and documented so that we can move to parsing SPARQL, SQL, and whatever other query languages we'd like to support into that representation directly instead of going through the analytical queries first, but I think that can wait a bit until we get all the FlureeQL stuff locked down.